flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializerSnapshot.java

```diff
@@ -0,0 +1,424 @@
+
+	/**
+	 * Transforms a {@link LinkedHashMap} with {@link TypeSerializerSnapshot}s as
+	 * the value to {@link TypeSerializer} as the value by restoring the snapshot.
+	 */
+	private static <K> LinkedHashMap<K, TypeSerializer<?>> restoreSerializers(LinkedHashMap<K, TypeSerializerSnapshot<?>> snapshotsMap) {
+		final LinkedHashMap<K, TypeSerializer<?>> restoredSerializersMap = new LinkedHashMap<>(snapshotsMap.size());
+		snapshotsMap.forEach((key, snapshot) -> restoredSerializersMap.put(key, snapshot.restoreSerializer()));
+		return restoredSerializersMap;
+	}
+
+	/**
+	 * Transforms the subclass serializer registry structure, {@code LinkedHashMap<Class<?>, TypeSerializer<?>>}
+	 * to 2 separate structures: a map containing with registered classes as key and their corresponding ids (order
+	 * in the original map) as value, as well as a separate array of the corresponding subclass serializers.
+	 */
+	@SuppressWarnings("unchecked")
+	private static Tuple2<LinkedHashMap<Class<?>, Integer>, TypeSerializer<Object>[]> decomposeSubclassSerializerRegistry(
+		LinkedHashMap<Class<?>, TypeSerializer<?>> subclassSerializerRegistry) {
+
+		final LinkedHashMap<Class<?>, Integer> subclassIds = new LinkedHashMap<>(subclassSerializerRegistry.size());
+		final TypeSerializer[] subclassSerializers = new TypeSerializer[subclassSerializerRegistry.size()];
+
+		subclassSerializerRegistry.forEach((registeredSubclassClass, serializer) -> {
+			int id = subclassIds.size();
+			subclassIds.put(registeredSubclassClass, id);
+			subclassSerializers[id] = serializer;
+		});
+
+		return Tuple2.of(subclassIds, subclassSerializers);
+	}
+
+	/**
+	 * Finds which Pojo fields exists both in the new {@link PojoSerializer} as well as in the previous one
+	 * (represented by this snapshot), and returns an {@link CompositeTypeSerializerUtil.IntermediateCompatibilityResult}
+	 * of the serializers of those preexisting fields.
+	 */
+	private static <T> CompositeTypeSerializerUtil.IntermediateCompatibilityResult<T> getCompatibilityOfPreExistingFields(
+			PojoSerializer<T> newPojoSerializer,
+			LinkedOptionalMap<Field, TypeSerializerSnapshot<?>> fieldSerializerSnapshots) {
+
+		// the present entries dictates the preexisting fields, because removed fields would be
+		// represented as absent keys in the optional map.
+		final Set<LinkedOptionalMap.KeyValue<Field, TypeSerializerSnapshot<?>>> presentFieldSnapshots =
+			fieldSerializerSnapshots.getPresentEntries();
+
+		final ArrayList<TypeSerializerSnapshot<?>> associatedFieldSerializerSnapshots = new ArrayList<>(presentFieldSnapshots.size());
+		final ArrayList<TypeSerializer<?>> associatedNewFieldSerializers = new ArrayList<>(presentFieldSnapshots.size());
+
+		for (LinkedOptionalMap.KeyValue<Field, TypeSerializerSnapshot<?>> presentFieldEntry : presentFieldSnapshots) {
+			TypeSerializer<?> associatedNewFieldSerializer = newPojoSerializer.getFieldSerializer(presentFieldEntry.getKey());
```

Comment:
this basically has O(n^2) performance, with n being the number of fields, since the `getFieldSerializer` method does a linear scan over the array of fields in the `PojoSerializer`. It is also like this in the current master branch.
Not sure if it is super critical, since this is only a one-time thing during restore.

To improve that, we essentially have to build an index in the `PojoSerializer` of the fields and their order.
@igalshilman what do you think about this?

Reply:
Really depends on the expected value of `n`, what would that be in practice 8,32,100?

I wouldn't add an index to the `PojoSerializer` but rather expose the fields (maybe with a package private visibility)
and build the index where it is needed. IMHO adding/removing fields right now should be carefully considered, due to backwards compatibility in respect with Java serialization, lazy initialization, serializer duplication.

Reply:
Building an index within the `getCompatibilityOfPreExistingFields` sounds like a good idea. I'll add that.

Reply:
Just realized there's also already a JIRA for this created before:
https://issues.apache.org/jira/browse/FLINK-6768